docs/shapes.md

```diff
@@ -0,0 +1,193 @@
+alive unused shapes. The pointers have been omitted, so the diagrams are clearer (too many arrows).
+
+Ok, now let us define a property `z` instead of `y`:
+
+```js
+o2.z = 300;
+```
+
+The following changes accure to the shape tree:
+
+```mermaid
+flowchart TD
+    Root(S0: Root Shape\n\nPrototype: None) --->| Property Count: 0 | PropertyTable[PropertyTable\n\nx: Slot 0, writable, configurable, enumerable\ny: Slot 1, has_get, has_set, configurable, enumerable]
+
+    ObjectPrototype(S1: Prototype Shape\n\nPrototype: Object.prototype) -->| Property Count: 0 |PropertyTable
+    ObjectPrototype -->|Previous| Root
+
+    InsertX(S2: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 1 | PropertyTable
+    InsertX -->|Previous| ObjectPrototype
+
+    InsertY(S3: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 2 | PropertyTable
+    InsertY -->|Previous| InsertX
+
+    InsertZ(S4: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 2 | PropertyTableFork[PropertyTable\n\nx: Slot 0, writable, configurable, enumerable\nz: Slot 1, writable, configurable, enumerable]
+    InsertZ -->|Previous| InsertX
+
+    O(Object o\n\nElement 0: JsValue: 100\nElement 1: JsValue: func\nElement 2: JsValue: func) --> InsertY
+    O2(Object o2\n\nElement 0: JsValue: 200\nElement 1: JsValue: 300) --> InsertZ
+```
+
+Now `o2` has `S4` shape. We can also see that `PropertyTable` has been forked, because we can no longer add a property add position `1`.
+
+What would happen if we wanted to delete a property `x` from object `o`:
+
+```js
+delete o.x;
+```
+
+```mermaid
+flowchart TD
+    Root(S0: Root Shape\n\nPrototype: None) --->| Property Count: 0 | PropertyTable[PropertyTable\n\nx: Slot 0, writable, configurable, enumerable\ny: Slot 1, has_get, has_set, configurable, enumerable]
+
+    ObjectPrototype(S1: Prototype Shape\n\nPrototype: Object.prototype) -->| Property Count: 0 |PropertyTable
+    ObjectPrototype -->|Previous| Root
+
+    InsertX(S2: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 1 | PropertyTable
+    InsertX -->|Previous| ObjectPrototype
+
+    InsertY(S3: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 2 | PropertyTable
+    InsertY -->|Previous| InsertX
+
+    InsertYNew(S3: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 1 |PropertyTableNew[PropertyTable\n\ny: Slot 0, has_get, has_set, configurable, enumerable]
+    InsertYNew -->|Previous| InsertX
+
+    O(Object o\n\nElement 0: JsValue: func\nElement 1: JsValue: func) --> InsertYNew
+```
+
+**NOTE:**: `o2` has been omitted and the it's shape.
+
+When a deletion happens, we find the node in the chain where we added the property, and get it's parent (`base`),
+we also remember that what transitions happened after the property insertion, then we apply them
+one by one until we construct the chain and return the last shape in that chain.
```

Comment:
I think this last one needs some tweaking, o2 has been omitted but its shape is still there, so it looks like a dangling shape with nothing referencing it.